Scrap Piles: Phase 1

NanakoAC opened the pull request:
My fixing-up of this much beleaguered PR
Scrap was originally a project being ported from elsewhere, its been passed around quite a number of times. It brings lots of content and useful systems, but it's a bit lacking in purpose and has some design flaws.

For this PR, i've chosen to only activate the parts which currently make sense, the rest is there but not really accessible yet. Implementing the other stuff will require some redesigning, bugfixing and mapping. 
I've also done a couple ancillary features which combine well.

So here's what is actually in this PR:

Junk piles:
Large searchable piles of random stuff, with dynamically generated sprites based on cobbling together many icons into a rather convincing pile. Also comes with six new SFX i made, thanks to various sources on freesound.org
These piles come with various themes. Some are just generic garbage, some are industrial, or medical, or full of broken science and tech. 

Searching can initially be done by hand but you only see part of its contents that way. To see it all, you need to use a shovel to dig through the pile and reveal more things. Many junk piles also have a large dense object hidden at the bottom, which you can only get to by clearing out the whole pile on top. You can examine the pile to see that such a thing is hidden, but you can't tell what it is until you uncover it.

Generally, the quality of stuff in these piles is low, but there's quite a bit of variety, and there's always the chance of a rare and powerful thing hidden in amongst the trash. 

Junk piles spawn all over maintenance, and are added to both the junk and structures random lists. In addition several pre-mapped junkpiles are placed in disposals areas, prison, spider and roach pits, and in maintenance near cryo, to provide a good first impression



Old items:
Implements a new make_old proc for atoms, which gives the item a decayed state. Many things turn brown or grey and get prefixes like "worn" or "rusty". Food becomes poisonous, clothing gets its armor stats and slowdown altered, grenades have their fuse timers randomised, etc.
Generally this proc turns any thing into a broken version of that thing. Many "old" items are found in these junk piles and generally scattered

This proc will be useful in future for derelict content, or also for things like temporal anomalies.


Random Mechas:
Mechas are now added to random maintenance structures. They are generally pretty rare. They can also be found buried under scrap piles, the possibility of which is a major incentive to dig them up.
Mechas implement the make_old proc with a set of complex malfunctions that i coded years ago and ported from aurora. So they can come in various states of repair, have all kinds of internal damage and misconfigured systems.  There's a very significant chance that they'll just be utterly broken mecha wreckage

Minor tweaks:
Added more closets to random lists, and a rare closets list too
Technomancers now start with an "almost full" toolbelt, has a 70% chance to contain each tool. To encourage a bit more scarcity and scavenging

-------------------------------------------------------------------------------------------------------
Now, what's missing. Here's a list of things that were part of this project, but are currently NOT available, and my explanation for each.

Scrap: Unrefined and Refined
Unrefined scrap is supposed to be dug out of the junk piles, and fed into a recycling machine to create refined scrap. The problem is, that there's currently no purpose for it. All you can do is burn it for fuel. It would be an extraneous and pointless mechanic to have currently.

But it has potential, i'd like to reactivate it in a future PR when we think of better purposes for it. Maybe refining materials or useful parts from it, using it in crafting recipes, etc


Scrap Machines:
There are quite a few machines in this code which aren't implemented yet. They would need to be mapped in, but also they don't have much purpose:
Pileripper: Supposed to be a quick way to search through a pile, but seems to be completely broken, it also can't move so you can't get it to piles. I'm not sure what the point of this one was.
I'm told there's supposed to be a mecha equipment to scoop up whole piles of junk and carry them to a recycling facility, so maybe thats something to add
Recycler: Turns unrefined scrap into refined scrap, works perfectly, but there's no purpose for the refined scrap atm
Stacker: Just stacks refined scrap, like the machine in mining. Probably doesnt need to exist, the mining machine could do the job
Burner: A pacman generator that runs on refined scrap. Not really enough of a purpose
Scrap Beacon: An utterly insane machine that causes junk piles to rain down on surrounding tiles. It seems to be intended for use on a planetside map, and doesn't make much sense on a ship yet.

NanakoAC commented:
> the point was that when the scrap is piled on you don't see the big item, but as you dig through you notice something.

Its not clear enough to the users, the description is added to provide people an incentive to search through the scrap pile right from the start
otherwise people would look at it once, see trash, and move on

LudwigVonChesterfield commented:
People should shovel the pile at least once, and then see what it has to offer, instead of glancing around.

Scrap piles shouldn't be treasuries, yet you really want to turn them into such.

NanakoAC commented:
> 
> 
> the belt will lose some of it's equipment due to oldification anyway... also this should be in Initialize, not New

This is for technmomancer starting equipment, not for oldified stuff. It's only a tangentially related change but one i tossed in here anyways. Generally reducing the quantity of free, perfect-condition normal tools around, so theres more incentive to craft and scavenge

RandomColonist commented:
Issues i found while testing:
1. Weathered fueltank that i found was 1000u full of fuel. Regular tank is 500, you know. It shouldn't be that full. Same with watertank exept i found 1500u tank.
2. 90% of items in tunnels overall NOT old. Only 10% is actually old. This should be reversed.
3. We have "Old closets" already. It should be integrated in new mechanics.
4. All of scrap piles has WAY TOO MUCH steel rods. That much that i can make one and a half full steel stack out of one pile. It should come in 1-5 rods quantity, not 10-30, and it should be less common.
5. If it's not already a thing, add significant chanse for any locker in maint become old. But that must not mean that ALL of items it contain is old. Maybe 90% or so. If it's already a thing - please, consider to raise this chance. I found zero old closets excluding excavated from piles and old "old closets".

LudwigVonChesterfield commented:
And Nanako was telling me to nerf oldification even further)

NanakoAC commented:
Alright, i pushed a commit with lots of fixes.

Points 1-5, fixed or improved
6 and 7 are working as intended, no changes
Fixed up various strings for 8
improved 9
10 is ehh, its fine
added ability to wash clothes for 11
12 is still an issue
13 is a rare result, but working as intended. the armor values on most of those will be shot to hell

RandomColonist commented:
Also, if you want to lower quality of tools on map and on character - removing random tools from technomancer belt will do literally nothing. There is multiple tool vending machines where you can buy tools really cheap. Do i need to mention that aurluxary tool storage (where ASSistants spawn) has 120 steel, 120 glass, 120 plastic and disc with all basic tools roundstart? My suggestions:

1. Outright remove Asters Basic Tools Pack from assistant room, leave only Misc Pack. You can already buy it or find it in tunnels with great chance. Also you might want to lower starting resources. Free autholathe =/= free resources. Because what they do is stealing metal and glass from it.

2. Rise price for tools in vendors and randomize quantity OR remove multiple vendors/restrict it by access (move them to engineering storage or something like that so engineers will decide do they want guild making profit from this vendors - more cooperation possibilities).

3. Replace some percent of regular tools with makeshift variants in tunnels. Like, in addition to tool being old it can just became makeshift. And the worst variant will be makeshift old tools. Still better than nothing.

4. About advanced tools - their balance is weird. You can craft a bunch for small price if you have disc, but if you don't - only scavenging left. So what i suggest is giving engineers a bunch of advanced tools. Especially Exultant. Or just give him disc with them, atlease. Cause it's logical for them to have it, they are involved in engineer-ish things more than anyone. RCD/Hatton is sure good things, but sometimes all i need as exultant is box of nice brand new advanced instruments.

5. Following same logic, slightly rebalance content of discs. For example, basic tools pack don't need multitool because multitool isn't really basical, it more belongs in vendors and advanced tools disc (just my opinion, discussable, i can explain it). Same goes for flamer. But this disc also lacks wire coils and simple tape. And advanced tools can use more advanced tools like flex tape, munchkin and so on.
6. Following same logic [2], add some advanced tools to science. I already mentioned that some tools can't be obtained besides maint looting. Maybe add it for a fair prise to protholathe designs? Like, you can spend fuckton of advanced resources to make advanced chainsaw/munchkin if you want it (and then sell it ofc).

7. Following same logic [3], add another toolset to uplink that will qualify like "advanced" and will cost twice or more of "cheap" (current) toolset. Make it also come in belt instead of unwieldy box. So you can pay 2 (i don't really remember) crystals for simple boxed set of tools or you can pay double and more for advanced brand new set of tools coming in the belt (with gloves too ofc). I don't think i need to explain that. Also you might want to add "disc" cathegory to uplink, but that's slightly unrelated.

I do realize that this is too much and unrelated, so i'l make separate suggestion and you can just anwer there or straight ignore it.

RandomColonist commented:
Here it is: #2859.

NanakoAC commented:
all these ideas are quite beyond scope, and not something to be done in one huge PR
in general, they should be filed and considered individually

kevinz000 commented on code/modules/scrap/scrap_beacon.dm:
```diff
@@ -0,0 +1,48 @@
+
+	active = TRUE
+	icon_state = "beacon1"
+	sleep(30)
```
Timers?

NanakoAC replied:
sleeps aren't necessarily awful, they're practical for small, simple things like this.

In any case the scrap beacon will not be active content for now

kevinz000 commented on code/modules/scrap/scrap_beacon.dm:
```diff
@@ -0,0 +1,48 @@
+	var/last_summon = -3000
+	var/active = 0
+
+/obj/structure/scrap_beacon/attack_hand(mob/user)
```
No ide aif you guys are going to ever port it but if you ever want to port component signals, stuff like attack_hand will need to call ..() for them to work right

kevinz000 commented on code/modules/scrap/scrap_beacon.dm:
```diff
@@ -0,0 +1,48 @@
+		new /obj/effect/falling_effect(newloc, /obj/random/scrap/moderate_weighted)
+	active = FALSE
+	icon_state = "beacon0"
+	return
```
unnecessary return :)

kevinz000 replied:
trailing return

kevinz000 commented on code/modules/scrap/scrap_burner.dm:
```diff
@@ -0,0 +1,23 @@
+	time_per_fuel_unit = 80
+
+/obj/machinery/power/port_gen/pacman/scrap/overheat()
+		explosion(loc, 1, 5, 2, -1)
```
uhhhh indent?

kevinz000 replied:
also what's with explosions often having "-1" as an argument, pretty sure -1 is just 0/null? Or am I wrong?

NanakoAC replied:
With explosions, a radius of 1 affects all tiles adjacent to you.
A radius of 0 affects only your own tile
a radius of -1 does nothing, which is the desired behaviour there

kevinz000 commented on code/modules/scrap/scrap_pileripper.dm:
```diff
@@ -0,0 +1,165 @@
+	if(default_part_replacement(I, user))
+		return
+
+	..()
```
return ..()?

kevinz000 replied:
return ..()

kevinz000 commented on code/modules/scrap/scrap_pileripper.dm:
```diff
@@ -0,0 +1,165 @@
+		if(safety_mode)
+			safety_mode = FALSE
+			update_icon()
+		playsound(loc, "sparks", 75, 1, -1)
```
pretty sure this should be src rather than loc but I'm not entirely sure if playsound here works like /tg/'s which does recommend people use src rather than its loc

kevinz000 commented on code/modules/scrap/scrap_refinery.dm:
```diff
@@ -0,0 +1,169 @@
+	desc = "A large crushing machine which is used to grind lumps of trash down; there are lights on the side of it."
+	icon = 'icons/obj/recycling.dmi'
+	icon_state = "grinder-o0"
+	layer = MOB_LAYER + 1 // Overhead
```
ABOVE_MOB_LAYER?

kevinz000 replied:
ABOVE_MOB_LAYER

kevinz000 commented on code/modules/scrap/scrap_refinery.dm:
```diff
@@ -0,0 +1,169 @@
+	icon_state = icon_name + "[is_powered]" + "[(blood ? "bld" : "")]" // add the blood tag at the end
+
+// This is purely for admin possession !FUN!.
+/obj/machinery/recycler/Bump(atom/movable/AM)
```
is Bumped() not usually called on base /atom/Bump()?

kevinz000 commented on code/modules/scrap/scrap_pileripper.dm:
```diff
@@ -0,0 +1,165 @@
+	desc = "This machine rips everything in front of it apart."
+	icon = 'icons/obj/structures/scrap/recycling.dmi'
+	icon_state = "grinder-b0"
+	layer = MOB_LAYER + 1 // Overhead
```
ABOVE_MOB_LAYER?

kevinz000 commented on code/modules/scrap/oldificator.dm:
```diff
@@ -0,0 +1,430 @@
+			//System settings will be randomly configured
+				var/P
+				if (prob(15))
+					cell.rigged = 1//Powercell will explode if you use it
```
TRUE?

kevinz000 commented on code/modules/scrap/oldificator.dm:
```diff
@@ -0,0 +1,430 @@
+				var/P
+				if (prob(15))
+					cell.rigged = 1//Powercell will explode if you use it
+				else if (prob(50))//Remove cell
```
qdel cell after?

kevinz000 commented on code/modules/scrap/oldificator.dm:
```diff
@@ -0,0 +1,430 @@
+			//The exosuit is wrecked. Spawns a wreckage object instead of a suit
+				//Set the noexplode var so it doesn't explode, then just qdel it
+				//The destroy proc handles wreckage generation
+				noexplode = 1
```
TRUE?

kevinz000 commented on code/modules/scrap/oldificator.dm:
```diff
@@ -0,0 +1,430 @@
+/obj/item/weapon/aiModule/broken/transmitInstructions(mob/living/silicon/ai/target, mob/sender)
+	..()
+	IonStorm(0)
+	explosion(sender.loc, 1, 1, 1, 3)
```
sender will be possibly qdel'd by the time drop_from_inventory is called if this gibs the user.

kevinz000 commented on code/modules/scrap/oldificator.dm:
```diff
@@ -0,0 +1,430 @@
+	return INITIALIZE_HINT_QDEL
+
+/obj/machinery/broken/Destroy()
+	contents.Cut()
```
forceMove contents to nullspace instead? Or qdel contents?

LethalGhost replied:
That was igonored?
:C

kevinz000 commented on code/modules/scrap/scrap.dm:
```diff
@@ -0,0 +1,497 @@
+	name = "This thing shoots scrap everywhere with a delay"
+	desc = "no data"
+	invisibility = 101
+	anchored = 1
```
TRUE

kevinz000 commented on code/modules/scrap/scrap.dm:
```diff
@@ -0,0 +1,497 @@
+	desc = "no data"
+	invisibility = 101
+	anchored = 1
+	density = 0
```
FALSE

kevinz000 commented on code/modules/scrap/scrap.dm:
```diff
@@ -0,0 +1,497 @@
+			if(H.shoes)
+				return
+
+			M << SPAN_DANGER("You step on \the [src]!")
```
to_chat

kevinz000 commented on code/modules/scrap/scrap.dm:
```diff
@@ -0,0 +1,497 @@
+	try_make_loot()
+	loot.open(user)
+	playsound(src, "rummage", 50, 1)
+	..()
```
. = ..()?

kevinz000 commented on code/modules/scrap/scrap.dm:
```diff
@@ -0,0 +1,497 @@
+
+/obj/structure/scrap/attack_generic(mob/user)
+	loot.open(user)
+	..()
```
return ..()?

kevinz000 commented on code/modules/scrap/scrap_beacon.dm:
```diff
@@ -0,0 +1,48 @@
+
+	active = TRUE
+	icon_state = "beacon1"
+	sleep(30)
```
timer?

TorinoFermic commented:
If I don't pick something out of a pile then I shovel this pile until this pile ceases to exist, all items I know that is there is gone.

NanakoAC commented:
I'm gonna declare the biohazard symbol out of scope for this project.
It is ugly and could do with work, but nothing functional requires that change to be in this PR. Its grown too bloated already and feature creep needs to stop. I'll fix the remaining things and make a commit shortly

NanakoAC commented:
New commit pushed.
Fixed paperbins
fixed several objects with missing icons
tweaked washing machine interactions a lot more

the biohazard symbol will not be fixed in this PR, that's a task for another one, and it can even be done independantly

LethalGhost commented on code/game/machinery/washing_machine.dm:
```diff
@@ -103,65 +131,26 @@
-		..()
+		var/item_allowed = FALSE
+		for (var/a in allowed_types)
+			if (istype(W, a))
```
`is_type_in_list(W, allowed_types)`

LethalGhost commented on code/game/machinery/washing_machine.dm:
```diff
@@ -103,65 +131,26 @@
+			if(contents.len < 10)
+				if ( state in list(1, 3) )
+					user.drop_item()
+					W.loc = src
```
`user.drop_item()` - drop item from active hand not drop item what was used.
`user.unEquip(W, src)` or `user.drop_from_inventory(W, src)` is much more preffered. Also `forceMove(...)` is missed

NanakoAC replied:
forceMove shouldnt be needed if i'm using one of those two, right?

LethalGhost commented on code/game/mecha/mecha.dm:
```diff
@@ -135,7 +136,7 @@
 
-	if(prob(30))
+	if(prob(30) && !noexplode)
 		explosion(get_turf(loc), 0, 0, 1, 3)
```
Actually it's better to move explosion code out of `Destroy(..)` proc. That shouldn't be here.

NanakoAC replied:
too much refactoring for now, can be done in a future mecha overhaul

LethalGhost commented on code/game/mecha/mecha.dm:
```diff
@@ -2159,3 +2160,48 @@ assassination method if you time it right*/
+	if (prob(probability))
+		radio.listening = !radio.listening
+	if (prob(probability))
+		radio.set_frequency(rand(1200,1600))
```
It's better to use defines for top/lower accessable frequencies.

LethalGhost commented on code/game/objects/items/weapons/storage/belt.dm:
```diff
@@ -57,6 +57,21 @@
+	if (prob(65))
+		new /obj/item/weapon/tool/wirecutters(src)
+	if (prob(65))
+		new /obj/item/stack/cable_coil(src,30,pick("red","yellow","orange"))
```
Isn't there exist predefines cable stuck with random color?

LethalGhost commented on code/game/objects/random/cloth.dm:
```diff
@@ -0,0 +1,180 @@
+	desc = "This is a random mask."
+
+/obj/random/cloth/masks/item_to_spawn()
+	return pick(subtypesof(/obj/item/clothing/mask/))
```
There exist few purevirtual types of masks (like one with AI vision). Also cigaretes is masks too (may be not a problem probably should be mentoined in comment here).

NanakoAC replied:
I checked out the mask in question, thats pretty cool. I'm tempted to allow it and just say that kind of thing is a perk

however, it throws runtimes. so i'll exclude it

LethalGhost commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -31,48 +31,59 @@
+					P = new /obj/item/weapon/paper/carbon
+
+			P.loc = user.loc
+			user.put_in_hands(P)
```
`P.loc = user.loc` already handles in `put_in_hands(...)` with proper calls like `forceMove(...)`

LethalGhost commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -31,48 +31,59 @@
+	if (usr == over_object)
+		if(pre_pickup(usr))
+			pickup(usr)
+			return TRUE
```
One can teleport paper_bin with that code?

NanakoAC commented:
I addressed everything that was posted before.

Thanks for reviewing, but you're a bit late :(  these comments should have been posted before merging, its been up over a week x
